source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1260,6 +1289,12 @@ void ClusterManagerImpl::postThreadLocalClusterUpdate(ClusterManagerCluster& cm_
       }
     }
   });
+
+  // By this time, the main thread has received the cluster initialization update, so we can start
+  // the ADS mux if the ADS mux is dependent on this cluster's initialization.
+  if (cm_cluster.requiredForAds()) {
```

Comment:
I think `postThreadLocalClusterUpdate` can be called more than once. In that case, IIUC, `ads_mux_->start()` may be initiated more than once, no?
I suggest also adding an assertion on the `start()` that it is only called once, and consider adding an integration test that updates the endpoints of the ADS cluster.